models/sessions.sql

```diff
@@ -1,11 +1,20 @@
 {{
     config(
-        materialized='view',
-        unique_key='full_session_id'
+        materialized='incremental',
+        unique_key='full_session_id',
+        partition_by={
+            "field": "updated_time",
+            "data_type": "timestamp",
+            "granularity": "day"
+        },
+        sort=['updated_time', 'start_time', 'end_time'],
+        dist=['device_id', 'user_id'],
+        cluster_by=['device_id', 'user_id'],
```

Comment:
How will this configuration would handle partial sessions? Imagine we compute aggregates for a session that has incomplete data, then `dbt run` after the session data is complete. Using the `append` strategy, we'll have a duplicate record that will need to be re-aggregated in some higher-level model. Using the `merge` strategy, we'll simply have incorrect data as the original record will be replaced.

This is the edge case that blocked us from releasing these as incrementals in the first version of this package, and it also applies to the users model. I'd love to know a) whether this configuration handles incomplete data correctly and b) if not, how we might re-approach to get the best of both worlds.

Reply:
@craigrmccown here's my thought pattern, but let me know if you see something flawed:
1. `stg_events__all` is the entire list of events that gets incrementally merged. We should insert new events, and update events where we already have the same ID.
2. `int_sessions` is a view built on `stg_events_all` so `int_sessions` would always be accurate, assuming `stg_events__all` is accurate. `updated_time` here should be accurate, _this is important_.
3. `sessions` is built from `int_sessions`. Rows with the same `full_session_id` will be replaced, new rows will be appended. Since `sessions` is an incremental table and we're only considering rows from `int_sessions` that have an `updated_time` > `max(updated_time)` in `sessions`. I think any row that should be updated / inserted will be returned here.

While recent session aggregates may change while we are receiving new data, I _think_ it should be accurate.

tldr: Since we are specifying a unique key in the `sessions` table and incrementally merging rows based on `updated_time` from `int_sessions` (a view), I _think_ the `sessions` aggregates will be accurate.

> ... as the original record will be replaced.
Yes, but it's replacement should be all-encompassing.

Reply:
@craigrmccown the main purpose for materializing the events staging table is to add proper indexing, partitioning and clustering. Ideally we should do this in the initial sync to warehouses.

This improvement makes the view (`int_sessions`) execution more efficient (which is only calculated once at dbt run time). The result is then materialized into the `sessions` table which has a similar improvement to it's setup.

If the consumer is primarily querying `sessions` or `users` tables this setup is better because the data is optimized for querying and we would rarely need to scan the entire data set (what we were doing previously).

Perhaps it's more accurate to say that we are now paying the expense once (during dbt run) rather than on every query.

I should also highlight the downside which is: if the data is _not_ queried, this is likely more costly. Before, `dbt run` was practically free, but every query would incur a significant compute cost (basically scanning the events table). Now, `dbt run` is not free (we pay for the incremental storage / compute), but query compute costs should be lower when using the `stg_events__all`, `sessions` or `users` tables.

Maybe it makes the most sense to explain the tradeoffs of the two strategies to the user, and allow them to specify a strategy via `vars`. I'm not sure the impact this would have on the maintainability of this package, but it's an option.